src/rust/engine/src/scheduler.rs

```diff
@@ -26,33 +28,83 @@ use ui::EngineDisplay;
 /// Both Scheduler and Session are exposed to python and expected to be used by multiple threads, so
 /// they use internal mutability in order to avoid exposing locks to callers.
 ///
-pub struct Session {
+struct InnerSession {
   // The total size of the graph at Session-creation time.
   preceding_graph_size: usize,
   // The set of roots that have been requested within this session.
   roots: Mutex<HashSet<Root>>,
   // If enabled, the display that will render the progress of the V2 engine.
   display: Option<Mutex<EngineDisplay>>,
+  // A place to store info about workunits in rust part
+  workunits: Option<Mutex<Vec<WorkUnit>>>,
```

Comment:
I don't think making this an `Option` is necessary if you conditionally disable in `Node::run` as mentioned above.